components/tel/Core/Src/freertos.c

```diff
@@ -0,0 +1,462 @@
+  /* USER CODE END RTOS_EVENTS */
+
+}
+
+/* USER CODE BEGIN Header_StartDefaultTask */
+/**
+  * @brief  Function implementing the defaultTask thread.
+  * @param  argument: Not used
+  * @retval None
+  */
+/* USER CODE END Header_StartDefaultTask */
+void StartDefaultTask(void *argument)
+{
+  /* USER CODE BEGIN StartDefaultTask */
+  /* Infinite loop */
+  for(;;)
+  {
+    osDelay(1);
+  }
+  /* USER CODE END StartDefaultTask */
+}
+
+/* Private application code --------------------------------------------------*/
+/* USER CODE BEGIN Application */
+
+__NO_RETURN void kernelLEDTask (void *argument) {
+  osKernelState_t kernel_status;
+
+  while (1) {
+    kernel_status = osKernelGetState();
+
+    if (kernel_status == osKernelRunning) {
+      HAL_GPIO_TogglePin(KERNEL_LED_GPIO_Port, KERNEL_LED_Pin);
+    }
+
+    osDelay(KERNEL_LED_DELAY);
+  }
+}
+
+__NO_RETURN void readCANTask(void *argument) {
+  static HAL_StatusTypeDef rx_status;
+  static CAN_msg_t current_can_message;
+
+  while (1) {
+    // wait for thread flags to be set
+    osThreadFlagsWait(CAN_READY, osFlagsWaitAll, osWaitForever);
+
+    if (HAL_CAN_GetRxFifoFillLevel(&hcan, CAN_RX_FIFO0) != 0) {
+      // there are multiple CAN IDs being passed through the filter, pull out the current message
+      rx_status = HAL_CAN_GetRxMessage(&hcan, CAN_RX_FIFO0, &can_rx_header, current_can_data);
+      HAL_GPIO_TogglePin(KERNEL_LED_GPIO_Port, KERNEL_LED_Pin);
+
+      // package into CAN_msg_t
+      current_can_message.header = can_rx_header;
+
+
+      for (uint8_t i=0; i<8; i++) {
+        current_can_message.data[i] = current_can_data[i];
+      }
+
+      // priority=0, timeout=0
+      osMessageQueuePut(canMessageQueueHandle, &current_can_message, 0U, 0U);
+    }
+
+    HAL_CAN_ActivateNotification(&hcan, CAN_IT_RX_FIFO0_MSG_PENDING);
+
+  }
+}
+
+__NO_RETURN void transmitMessageTask(void *argument) {
+  static CAN_msg_t can_message;
+  osStatus_t queue_status;
+
+  while (1) {
+    // retrieve CAN message from queue
+    queue_status = osMessageQueueGet(canMessageQueueHandle, &can_message, NULL, osWaitForever);
+
+    if (queue_status != osOK){
+      osThreadYield();
+    }
+
+    uint8_t can_buffer[21] = {0};
+
+    // TIMESTAMP: 8 ASCII characters
+    for (uint8_t i=0; i<8; i++) {
+      // send 'D' as placeholder
+        can_buffer[i] = 'D';
+    }
+
+    // CAN ID: 4 ASCII characters
+    uint8_t id_h = 0xFFUL & (can_message.header.StdId >> 8);
+    uint8_t id_l = 0xFFUL & (can_message.header.StdId);
+
+    can_buffer[8] = id_h;
+    can_buffer[9] = id_l;
+
+
+    // CAN DATA: 16 ASCII characters
+    for (uint8_t i=0; i<8; i++) {
+      // can_stream[2+i] = 0xFFUL & (can_message.data[i]);
+        can_buffer[i + 10]= can_message.data[i];
+    }
+
+    // CAN DATA LENGTH: 1 ASCII character
+    uint8_t length = "0123456789ABCDEF"[ can_message.header.DLC & 0xFUL];
+    can_buffer[18] = length;
+
+    // NEW LINE: 1 ASCII character
+    can_buffer[19] = '\n';
```

Comment:
Same as these, they could be macros or `CAN_BUFFER_LEN - 1`, or something else that makes sense